### PR TITLE
Add ansible for provisioning projects,users,groups

### PR DESCRIPTION
**Pull request by eggshell**
This adds the ability to provision our desired projects, groups and
users on the new cloud. It is a continuation of the cloud_bootstrap
role started in BonnyCI/hoist#166, and follows the spec contained in
BonnyCI/bonnyci.org#34.

Depends-On: BonnyCI/hoist#166
Resolves: BonnyCI/projman#134

### CODE REVIEW

File: create-users-projects.yml
Comment (omgjlk) [inline create-users-projects.yml]: If we're going to make this more of a bootstrap role, we should remove the users-projects tag off of the role itself, and instead rely on the tags inside the role.

Should probably re-name the play file too.

File: roles/cloud-bootstrap/tasks/create-users-projects.yml
Comment (omgjlk) [inline roles/cloud-bootstrap/tasks/create-users-projects.yml]: This should be a variable from the role defaults.

File: roles/cloud-bootstrap/tasks/create-users-projects.yml
Comment (omgjlk) [inline roles/cloud-bootstrap/tasks/create-users-projects.yml]: Same with this, a role variable.

File: roles/cloud-bootstrap/tasks/create-users-projects.yml
Comment (omgjlk) [inline roles/cloud-bootstrap/tasks/create-users-projects.yml]: Same.

File: roles/cloud-bootstrap/tasks/create-users-projects.yml
Comment (omgjlk) [inline roles/cloud-bootstrap/tasks/create-users-projects.yml]: variable

File: roles/cloud-bootstrap/tasks/create-users-projects.yml
Comment (omgjlk) [inline roles/cloud-bootstrap/tasks/create-users-projects.yml]: the task names should be more generic, such as the `production project` or some such. Other implementers of this stuff may not want to call it bonnyci.

File: roles/cloud-bootstrap/tasks/create-users-projects.yml
Comment (omgjlk) [inline roles/cloud-bootstrap/tasks/create-users-projects.yml]: this role should be `_member_`, but also should be a variable.

In fact, the existing role `users-projects` did this bit. Your commit sort of leaves that role an orphan. Could that code instead be used in this role?

File: roles/cloud-bootstrap/tasks/create-users-projects.yml
Comment (omgjlk) [inline roles/cloud-bootstrap/tasks/create-users-projects.yml]: These should be variables.

File: roles/cloud-bootstrap/tasks/create-users-projects.yml
Comment (omgjlk) [inline roles/cloud-bootstrap/tasks/create-users-projects.yml]: variable

File: roles/cloud-bootstrap/tasks/create-users-projects.yml
Comment (omgjlk) [inline roles/cloud-bootstrap/tasks/create-users-projects.yml]: It's preferable to use the `bool` filter to check for truthiness. Also, since this attribute might only be set WHEN a user is an admin, we have to handle the default case where it is not defined:

`when: item.isadmin | default('False') | bool`

File: roles/cloud-bootstrap/tasks/create-users-projects.yml
Comment (omgjlk) [inline roles/cloud-bootstrap/tasks/create-users-projects.yml]: variables.

File: roles/cloud-bootstrap/tasks/create-users-projects.yml
Comment (omgjlk) [inline roles/cloud-bootstrap/tasks/create-users-projects.yml]: all of these need variables.

File: roles/cloud-bootstrap/tasks/create-users-projects.yml
Comment (omgjlk) [inline roles/cloud-bootstrap/tasks/create-users-projects.yml]: With ansible 2.x, the argument to loops needs to be templated.

`with_items: {{ users }}`

File: roles/cloud-bootstrap/tasks/main.yml
Comment (omgjlk) [inline roles/cloud-bootstrap/tasks/main.yml]: since the previous tag in this role is more simplistic, `network`, we could be simple in this tag as well, `accounts`. or some such.

However, since the network bits require the users/projects in place, this new task set to create the users and projects needs to run before the network tasks.

**Review by omgjlk — CHANGES_REQUESTED** [review]
Comments in-line.

File: roles/cloud-bootstrap/tasks/main.yml
Comment (gandelman-a) [reply roles/cloud-bootstrap/tasks/main.yml]: So actually this is something we'll need to figure out once the network stuff merges... is how we create the bonnyci project here, and use it in network.yml when creating those resources. Currently sifting thru the os modules+shade+os-client-config to see if there's a way we can just specify it in the play, but it looks like we may need to keep multiple clouds.yaml entries for each project

Comment (gandelman-a) [reply roles/cloud-bootstrap/tasks/main.yml]: Actually my point looks easily resolved. Once we create the project, we can pass an auth parameter to the resources in network.yml, that specifies only the project and it will override in the auth provided the user running the play has access.

**Comment by anne-bonny** [comment]
Build failed.

- http://logs.bonnyci.com/check_github/BonnyCI/hoist/168/1487885728.63/bonnyci-run-check-multinode : FAILURE in 56s

File: roles/cloud-bootstrap/tasks/network.yml
Comment (gandelman-a) [inline roles/cloud-bootstrap/tasks/network.yml]: So in all of these os_* tasks, we need to specify the project we want to create them in with an auth: parameter, ie:
```
auth:
  project_name: bonnyci
```

Comment (eggshell) [reply roles/cloud-bootstrap/tasks/network.yml]: @gandelman-a cool. I'll switch up the order of `accounts.yml` and `network.yml` and add the auth parameters.

File: roles/cloud-bootstrap/tasks/create-users-projects.yml
Comment (eggshell) [reply roles/cloud-bootstrap/tasks/create-users-projects.yml]: @gandelman-a I used `secrets.nodepool.db_password` since I didn't see just a `password` var in the example secrets.

File: roles/cloud-bootstrap/tasks/create-users-projects.yml
Comment (eggshell) [reply roles/cloud-bootstrap/tasks/create-users-projects.yml]: @j2sol I think the `user-projects` role assumes a slightly different approach to our account provisioning, but I do agree that the code is extremely similar. Do you think we should remove that once this is merged and fully tested?

**Review by omgjlk — CHANGES_REQUESTED** [review]
getting better, still some problems.

File: create-users-projects.yml
Comment (omgjlk) [inline create-users-projects.yml]: This tag is inappropriate, as it applies the `accounts` tag to the entire role, and inside that role, both accounts _and_ networks are created. It'd be best to remove the tag here all together, and just rely on the tags inside the `main.yml` file within the role.

File: roles/cloud-bootstrap/tasks/accounts.yml
Comment (omgjlk) [inline roles/cloud-bootstrap/tasks/accounts.yml]: This will probably fail. `users.username` has no meaning at this point. I think you may want to do some concatenation for the three variable sets, or since that's awkward, just do one task that has the `prod_project` and `nodepool_project` items, and a second task that loops over `users` and uses `item.username`. No need to be really clever, just make it clear to understand.

**Review by omgjlk — CHANGES_REQUESTED** [review]
really close now! Just need to fix up tags and some of the loops, and then I think you can just delete all of the `users-projects` role as part of this change.

File: roles/cloud-bootstrap/tasks/network.yml
Comment (gandelman-a) [reply roles/cloud-bootstrap/tasks/network.yml:7]: Right but this will be run from an admin locally and may have a different project (their personal) set on the cloud their using.

**Comment by gandelman-a** [comment]
I think we still need the ``users-projects`` role, to manage the personal dev and admin accounts that will eventually be running this?  We could potentially role it into the ``cloud-bootstrap`` role/

**Comment by gandelman-a** [comment]
This LGTM, will let @j2sol look and change his negative to a positive.

**Review by ConsoleCatzirl — COMMENTED** [review]
:star: looks good to me, but someone else should approve

File: roles/cloud-bootstrap/tasks/accounts.yml
Comment (omgjlk) [inline roles/cloud-bootstrap/tasks/accounts.yml]: This needs to set a password, defaulting to a simple one, like in the previous role. It also needs to set the task to not update the password on subsequent runs.

```
     password: "{{ default_password }}"
     update_password: on_create
```

File: roles/cloud-bootstrap/tasks/accounts.yml
Comment (omgjlk) [inline roles/cloud-bootstrap/tasks/accounts.yml]: Should copy over the task that grants the user a role on their project, otherwise they won't be able to do anything.

Grab the task from https://github.com/BonnyCI/hoist/pull/168/files#diff-e0b563178351dc3acf8378b82947655eL18 and make sure any defaults used are ported over as well.

**Review by omgjlk — CHANGES_REQUESTED** [review]

**Review by mlangbehn — COMMENTED** [review]
Looks good to me. @j2sol, did you get a chance to look back over this yet?

**Review by omgjlk — APPROVED** [review]
Woo thanks for sticking it out on this one, lots of back and forth. Lets see how it blows up!

**Pull request merged**
